waspc/src/Wasp/SemanticVersion.hs

```diff
@@ -78,27 +106,74 @@ instance Semigroup Range where
+  case operator of
+    Equal -> version == compVersion
+    LessThan -> version < compVersion
+    LessThanOrEqual -> version <= compVersion
+    GreaterThan -> version > compVersion
+    GreaterThanOrEqual -> version >= compVersion
+
+comparatorSetToPrimitiveComparatorSet :: ComparatorSet -> PrimitiveComparatorSet
+comparatorSetToPrimitiveComparatorSet (ComparatorSet comps) =
+  foldl1 (<>) $ comparatorToPrimitiveComparatorSet <$> comps
+
+comparatorToPrimitiveComparatorSet :: Comparator -> PrimitiveComparatorSet
+comparatorToPrimitiveComparatorSet (ComparatorP comp) =
+  PrimitiveComparatorSet $ fromList [comp]
+comparatorToPrimitiveComparatorSet (ComparatorS comp) =
+  specialComparatorToPrimitiveComparatorSet comp
+
+specialComparatorToPrimitiveComparatorSet :: SpecialComparator -> PrimitiveComparatorSet
+specialComparatorToPrimitiveComparatorSet (Caret version) =
+  PrimitiveComparatorSet $
+    NE.fromList
+      [ PrimitiveComparator GreaterThanOrEqual version,
+        PrimitiveComparator LessThan (nextBreakingChangeVersion version)
+      ]
+
 nextBreakingChangeVersion :: Version -> Version
 nextBreakingChangeVersion version = case version of
   (Version 0 0 x) -> Version 0 0 (succ x)
   (Version 0 x _) -> Version 0 (succ x) 0
   (Version x _ _) -> Version (succ x) 0 0
 
-doesVersionSatisfyComparator :: Version -> Comparator -> Bool
-doesVersionSatisfyComparator version (Comparator operator compVersion) = case operator of
-  Equal -> version == compVersion
-  LessThanOrEqual -> version <= compVersion
-  BackwardsCompatibleWith -> version >= compVersion && version < nextBreakingChangeVersion compVersion
+-- Helper methods.
 
-doesVersionSatisfyComparatorSet :: Version -> ComparatorSet -> Bool
-doesVersionSatisfyComparatorSet version (ComparatorSet comps) = all (doesVersionSatisfyComparator version) comps
+(&) :: ComparatorSet -> ComparatorSet -> ComparatorSet
+(&) = (<>)
 
-isVersionInRange :: Version -> Range -> Bool
-isVersionInRange version (Range compSets) = any (doesVersionSatisfyComparatorSet version) compSets
+lt :: Version -> ComparatorSet
+lt = mkPrimCompSet LessThan
+
+lte :: Version -> ComparatorSet
+lte = mkPrimCompSet LessThanOrEqual
+
+gt :: Version -> ComparatorSet
+gt = mkPrimCompSet GreaterThan
+
+gte :: Version -> ComparatorSet
+gte = mkPrimCompSet GreaterThanOrEqual
+
+eq :: Version -> ComparatorSet
+eq = mkPrimCompSet Equal
 
-rangeFromVersion :: (Operator, Version) -> Range
-rangeFromVersion = Range . pure . ComparatorSet . pure . uncurry Comparator
+caret :: Version -> ComparatorSet
+caret = ComparatorSet . pure . ComparatorS . Caret
 
-rangeFromVersionsIntersection :: [(Operator, Version)] -> Range
-rangeFromVersionsIntersection [] = Range []
-rangeFromVersionsIntersection compPairs = Range $ pure $ ComparatorSet $ uncurry Comparator <$> NE.fromList compPairs
+mkPrimCompSet :: Operator -> Version -> ComparatorSet
+mkPrimCompSet op = ComparatorSet . pure . ComparatorP . PrimitiveComparator op
```

Comment:
Ha, good points, but this is a very local function, used only in the couple of functions above to make them more readable. So due to context I would argue that it being abbreviated is not an issue but a feature :D.
As for the name -> well yes, it does create ComparatorSet at the end, but it actually creates primitive comparator set that is wrapped into ComparatorSet which is a wrapper for these. So I think it is ok name for this small local situation. How would you call it otherwise?

Reply:
Oh, I see. Everything makes sense then.